dig.go

```diff
@@ -275,3 +307,85 @@ func detectCycles(n node, graph map[reflect.Type]node, path []reflect.Type, seen
 	}
 	return nil
 }
+
+// Param is embedded inside structs to opt those structs in as Dig parameter
+// objects.
+//
+// TODO usage docs
+type Param struct{}
+
+var _ parameterObject = Param{}
+
+// Param is the only instance of parameterObject.
+func (Param) parameterObject() {}
+
+// Users embed the Param struct to opt a struct in as a parameter object.
+// Param implements this interface so the struct into which Param is embedded
+// also implements this interface. This provides us an easy way to check if
+// something embeds Param without iterating through all its fields.
+type parameterObject interface {
+	parameterObject()
+}
+
+// Returns dependencies introduced by a parameter object.
+func getParameterDependencies(t reflect.Type) ([]reflect.Type, error) {
+	for t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+
+	var deps []reflect.Type
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		if f.PkgPath != "" {
+			continue // skip private fields
+		}
+
+		// Skip the embedded Param type.
+		if f.Anonymous && f.Type == _paramType {
+			continue
+		}
+
+		// THe user added a parameter object as a dependency. We don't recurse
+		// /yet/ so let's try to give an informative error message.
+		if f.Type.Implements(_parameterObjectType) {
```

Comment:
Interfaces with unexported methods may only be implemented by types defined in
that package. A type that implements a private interface may be embedded
inside another type (which may reside outside the package containing the
private interface). Because methods of embedded types are "inherited", the
type that embeds the other type now implements the private interface too.

That's exactly the behavior we want: Only types that embed `dig.Param`
implement `parameterObject`. Demonstration:

```go
// foo/foo.go
package foo

type A interface { private() }

func Matches(x interface{}) bool {
	_, ok := x.(A)
	return ok
}

// External types embedding this will match.
type EmbedMe struct{}

func (EmbedMe) private() {}
```

```go
// main.go
package main

import (
	"fmt"
	"tmp/foo"
)

type bad struct{}

func (bad) private() {}

type good struct{ foo.EmbedMe }

func main() {
	fmt.Println("bad:", foo.Matches(bad{}))
	fmt.Println("good:", foo.Matches(good{}))
	// Output:
	// bad: false
	// good: true
}
```

In your example, the other package doesn't define the interface, only the
method. And you embed it in a type inside the package where you define the
interface, so it somehow matches.

This will not work for types that simply implement `parameterObject()`.